src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -2922,7 +2924,7 @@ private LLVMValueRef GetFieldAddress(FieldDesc field, bool isStatic)
 
                 // Run static constructor if necessary
                 // If the type is non-BeforeFieldInit, this is handled before calling any methods on it
-                if (owningType.IsBeforeFieldInit || (!owningType.IsBeforeFieldInit && owningType != _thisType))
+                if (!cctorCalled && (owningType.IsBeforeFieldInit || (!owningType.IsBeforeFieldInit && owningType != _thisType)))
```

Comment:
I think we can factor this in a nicer way and keep this optimization for the thread static case as well (elide cctor check if we're trying to access a non-beforefieldinit static field on the same type as the currently compiled method).

What I would do is:
* Move the expression `owningType.IsBeforeFieldInit || (!owningType.IsBeforeFieldInit && owningType != _thisType)` up to the place where you initialize the new cctorCalled bool (delete that bool). Assign the value of the expression to a new variable (e.g. `needsCctorCheck`).
* Move the TriggerCctor call from here to the individual places where it's needed, under a `if (needsCctorCheck)` condition.
* Pass the value of `needsCctorCheck` to `TriggerCctorWithThreadStaticStorage`

For bonus points, I would also move the `_compilation.TypeSystemContext.HasLazyStaticConstructor(type)` check and make it part of `needsCctorCheck`. It's odd that we have a method named `TriggerCctor` that may not actually trigger anything.

Reply:
Thanks, thought I might get pulled up on that double out signature.  Addressed in lastest push.